src/transformers/modeling_tf_bart.py

```diff
@@ -0,0 +1,1215 @@
+                v = self.v_proj(key)
+        else:
+            k = self.k_proj(query)
+            v = self.v_proj(query)
+
+        q = self._shape(q, tgt_len, bsz)
+        if k is not None:
+            k = self._shape(k, -1, bsz)
+        if v is not None:
+            v = self._shape(v, -1, bsz)
+
+        if saved_state is not None:
+            k, v, key_padding_mask = self._use_saved_state(k, v, saved_state, key_padding_mask, static_kv, bsz)
+
+        # Update cache
+        layer_state[self.cache_key] = {
+            "prev_key": tf.reshape(k, (bsz, self.num_heads, -1, self.head_dim)),
+            "prev_value": tf.reshape(v, (bsz, self.num_heads, -1, self.head_dim)),
+            "prev_key_padding_mask": key_padding_mask if not static_kv else None,
+        }
+
+        assert k is not None
+        src_len = k.shape[1]
+        attn_weights = tf.matmul(q, k, transpose_b=True)
+
+        assert attn_weights.shape == (bsz * self.num_heads, tgt_len, src_len)
+
+        if attn_mask is not None:
+            assert attn_mask.dtype == tf.float32, f"expected dtype tf.float32 got {attn_mask.dtype}"
+            attn_weights = tf.reshape(attn_weights, (bsz, self.num_heads, tgt_len, src_len)) + attn_mask
+            attn_weights = tf.reshape(attn_weights, (bsz * self.num_heads, tgt_len, src_len))
+
+        if key_padding_mask is not None and key_padding_mask._rank() == 0:
+            key_padding_mask = None
+        assert key_padding_mask is None or key_padding_mask.shape[:2] == (
+            bsz,
+            src_len,
+        )
+        if key_padding_mask is not None:  # don't attend to padding symbols
+            attn_weights: tf.Tensor = tf.reshape(attn_weights, (bsz, self.num_heads, tgt_len, src_len))
+            if key_padding_mask.dtype == tf.bool:
+                key_padding_mask = tf.cast(key_padding_mask, attn_weights.dtype) * -1e9
+            extended_mask = tf.expand_dims(tf.expand_dims(key_padding_mask, 1), 2)
+            attn_weights = attn_weights + extended_mask
+            attn_weights = tf.reshape(attn_weights, (bsz * self.num_heads, tgt_len, src_len))
+
+        attn_weights = tf.nn.softmax(attn_weights, axis=-1)
+        attn_probs = tf.nn.dropout(attn_weights, rate=self.dropout if training else 0.0)
+
+        assert v is not None
+        attn_output = tf.matmul(attn_probs, v)
+        assert attn_output.shape == (bsz * self.num_heads, tgt_len, self.head_dim)
+        attn_output = tf.transpose(attn_output, perm=(1, 0, 2))
+        attn_output = tf.reshape(attn_output, (tgt_len, bsz, embed_dim))
+        attn_output = self.out_proj(attn_output)
+        attn_weights: tf.Tensor = tf.reshape(attn_weights, (bsz, self.num_heads, tgt_len, src_len))
+        return attn_output, attn_weights
+
+    def _use_saved_state(self, k, v, saved_state, key_padding_mask, static_kv, bsz):
+        # saved states are stored with shape (bsz, num_heads, seq_len, head_dim)
+        if "prev_key" in saved_state:
```

Comment:
What I don't really understand here is that if `saved_state` is not None => "prev_key" is always in `saved_state` (It's added a couple of lines above. So I think, this whole function has a lot of code that makes it hard to read, but could be simplified to like 4 lines: 

 ```
# key states
prev_key = tf.reshape(saved_state["prev_key"], (bsz * self.num_heads, -1, self.head_dim))
key_states = prev_key if static_kv else tf.concat([prev_key, k], axis=1)

# value states
prev_value = tf.reshape(saved_state["prev_value"], (bsz * self.num_heads, -1, self.head_dim))
value_states = prev_value if static_kv else tf.concat([prev_value, v], axis=1)

The assert's are all quite misleading I think. If this function is called  1) "prev_key" and "prev_value" have to be in `saved_state` and k and v have to be not `None` afterwards.